### PR TITLE
Implemented sinc function

### LINKED ISSUE #5051: Implement sinc function
```
SymPy should implement a sinc function with correct evaluation at x = -oo,
0 and oo.

See: http://groups.google.com/group/sympy/browse_thread/thread/9243051af53c786?hl=en
```
Original issue for #5051: http://code.google.com/p/sympy/issues/detail?id=1952
Original author: https://code.google.com/u/111502149103757882156/

### PR DESCRIPTION
**Pull request by leosartaj**
Implemented sinc class

Examples

```
>>> from sympy import sinc, oo, jn, Product, Symbol
>>> from sympy.abc import x
>>> sinc(x)
sinc(x)
```

Automated Evaluation

```
>>> sinc(0)
1
>>> sinc(oo)
0
```

Differentiation

```
>>> sinc(x).diff()
(x*cos(x) - sin(x))/x**2
```

Series Expansion

```
>>> sinc(x).series()
1 - x**2/6 + x**4/120 + O(x**6)
```

As zero'th order spherical Bessel Function

```
>>> sinc(x).rewrite(jn)
jn(0, x)
```

Product form

```
>>> k = Symbol('k', integer=True, positive=True)
>>> sinc(x).rewrite(Product)
Product(2**(-k)*cos(x), (k, 1, oo))
```

fixes #5051

@certik @asmeurer

### CODE REVIEW

File: sympy/functions/elementary/trigonometric.py [hunk 1]
```diff
@@ -1568,6 +1568,97 @@ def taylor_term(n, x, *previous_terms):
...
+        if arg.is_Number:
+            if arg is S.Zero:
+                return S.One
+            elif arg in [S.Infinity, -S.Infinity]:
+                return S.Zero
+            elif arg is S.NaN:
+                return S.NaN
+
+        if arg is S.ComplexInfinity:
+            return S.ComplexInfinity
```
Comment (aktech) [inline sympy/functions/elementary/trigonometric.py]: `sinc(zoo)` is probably indeterminate.

**Comment by aktech** [comment]
You could also add sine rewrite:
```python
def _eval_rewrite_as_sin(self, arg):
        return sin(arg)/arg
```

Otherwise, it looks good to me.

File: sympy/functions/elementary/trigonometric.py [hunk 2]
```diff
@@ -1568,6 +1568,97 @@ def taylor_term(n, x, *previous_terms):
...
+
+    Examples
+    ========
+
+    >>> from sympy import sinc, oo, jn, Product, Symbol
+    >>> from sympy.abc import x
+    >>> sinc(x)
+    sinc(x)
+
+    Automated Evaluation
```
Comment (aktech) [inline sympy/functions/elementary/trigonometric.py]: Automated Evaluation  --->  * Automated Evaluation 
(It would look better in docs)

File: sympy/functions/elementary/trigonometric.py [hunk 3]
```diff
@@ -1568,6 +1568,97 @@ def taylor_term(n, x, *previous_terms):
...
+
+    >>> sinc(0)
+    1
+    >>> sinc(oo)
+    0
+
+    Differentiation
```
Comment (aktech) [inline sympy/functions/elementary/trigonometric.py]: same

File: sympy/functions/elementary/trigonometric.py (hunk 1 above)
Comment (jksuom) [inline sympy/functions/elementary/trigonometric.py]: Yes, ``sinc`` has no limit at ``zoo`` ([an essential singularity](http://en.wikipedia.org/wiki/Essential_singularity)). I believe the value should be ``nan`` in SymPy.

**Comment by leosartaj** [comment]
I have made the changes @aktech @jksuom

**Comment by jksuom** [comment]
I am worried about this
```
In [3]: s = Symbol('s', zero=True)

In [6]: sin(s*pi)
Out[6]: 0

In [7]: sinc(s*pi)
Out[7]: 0
```

**Comment by jksuom** [comment]
References to ``Product(cos(arg)/2**k, (k, 1, S.Infinity))`` and ``Integral(sinc(t), (t, 0, z))`` would be useful.

**Comment by toolforger** [comment]
Am 09.04.2015 um 11:48 schrieb Kalevi Suominen:
> I am worried about this
> ```
> In [3]: s = Symbol('s', zero=True)
>
> In [6]: sin(s*pi)
> Out[6]: 0
>
> In [7]: sinc(s*pi)
> Out[7]: 0
> ```

Right - sinc's singularity at zero is removable, but the value should be 
`1` (or `NaN` if SymPy does not remove removable singularities, dunno 
SymPy's policy for that).

**Comment by leosartaj** [comment]
This should return 1

```
In [3]: s = Symbol('s', zero=True)

In [6]: sin(s*pi)
Out[6]: 0

In [7]: sinc(s*pi)
Out[7]: 1
```

I will fix this.

**Comment by leosartaj** [comment]
I have fixed the following

```
>>> n = Symbol('n', integer=True)
>>> sinc(n*pi)
sinc(n*pi)
>>> s = Symbol('s', zero=True)
>>> sinc(s*pi)
1
>>> k = Symbol('k', integer=True, nonzero=True)
>>> sinc(k*pi)
0
```

@jksuom

File: sympy/functions/elementary/trigonometric.py [hunk 4]
```diff
@@ -1568,6 +1568,100 @@ def taylor_term(n, x, *previous_terms):
...
+        if arg is S.ComplexInfinity:
+            return S.NaN
+
+        if arg.could_extract_minus_sign():
+            return cls(-arg)
+
+        pi_coeff = _pi_coeff(arg)
+        if pi_coeff is not None:
+            if pi_coeff.is_integer:
+                if arg.is_nonzero:
```
Comment (jksuom) [inline sympy/functions/elementary/trigonometric.py]: This is probably not necessary any more as ``arg.is_zero`` is tested above.

Comment (leosartaj) [inline sympy/functions/elementary/trigonometric.py]: This is required for the case

```
>>> n = Symbol('n', integer=True) # may or may not be zero
>>> sinc(n*pi)
sinc(n*pi)
```

Comment (jksuom) [inline sympy/functions/elementary/trigonometric.py]: Yes, that is right. Though ``pi_coeff.is_nonzero`` might be more suitable as the assumption mechanism would not need to check ``pi.is_nonzero``. 
(Perhaps a comment could be added here.)

**Comment by jksuom** [comment]
I think that references to the formulas with Product and Integral would be really needed.

File: sympy/functions/elementary/trigonometric.py (hunk 4 above)
Comment (leosartaj) [inline sympy/functions/elementary/trigonometric.py]: From the docstring of _pi_coeff, it returns in the range [0, 2].

```
>>> from sympy.functions.elementary.trigonometric import _pi_coeff
>>> _pi_coeff(10*pi)
0
```

This will cause problems in the cases, like
```
>>> sinc(10*pi)
sinc(10*pi)
```

I think checking arg is better here.

**Comment by leosartaj** [comment]
I already did add them

```
>>> sinc(x).rewrite(Product)
Product(2**(-k)*cos(x), (k, 1, oo))
>>> Si(x).rewrite(sinc)
Integral(sinc(t), (t, 0, x))
```

**Comment by jksuom** [comment]
I am sorry, perhaps I should have expressed myself more clearly. I meant references to the literature justifying these formulas.

**Comment by leosartaj** [comment]
Fine, I will add some references. Also, will it be appropriate to have the following functionality

```
>>> sinc(x).rewrite(Si)
Derivative(Si, x)
```

File: sympy/functions/elementary/trigonometric.py (hunk 4 above)
Comment (jksuom) [inline sympy/functions/elementary/trigonometric.py]: Ok. I forgot it was normalized. That is not so useful here as in the case of ``sin()``.

File: sympy/functions/elementary/trigonometric.py [hunk 5]
```diff
@@ -1568,6 +1568,100 @@ def taylor_term(n, x, *previous_terms):
...
+                    return S.Zero
+            elif (2*pi_coeff).is_integer:
+                return S.NegativeOne**(pi_coeff - S.Half) / arg
```
Comment (jksuom) [inline sympy/functions/elementary/trigonometric.py]: Should there be a test exercising this line?

**Comment by jksuom** [comment]
> Also, will it be appropriate to have the following functionality
>>> sinc(x).rewrite(Si)
Derivative(Si, x)

I doubt if a derivative would qualify as a rewrite. My knowledge of SymPy policy is insufficient.
@asmeurer ?

File: sympy/functions/elementary/trigonometric.py (hunk 5 above)
Comment (leosartaj) [inline sympy/functions/elementary/trigonometric.py]: I have added some tests, testing this piece of code.

**Comment by leosartaj** [comment]
I have improved the docs. Have a look @jksuom.

File: sympy/functions/elementary/trigonometric.py [hunk 6]
```diff
@@ -1568,6 +1568,101 @@ def taylor_term(n, x, *previous_terms):
...
+        from sympy.functions.special.bessel import jn
+        return jn(0, arg)
+
+    def _eval_rewrite_as_sin(self, arg):
+        return sin(arg) / arg
+
+    def _eval_rewrite_as_Product(self, arg):
+        from sympy import Product
+        k = Symbol('k', Dummy=True, integer=True, positive=True)
+        return Product(cos(arg)/2**k, (k, 1, S.Infinity))
```
Comment (jksuom) [inline sympy/functions/elementary/trigonometric.py]: I have had doubts about whether this product formula should be called a rewrite as it is an infinite product. There seems to be no comparable rewrite method in SymPy. Besides, there are other infinite product formulas for ``sinc`` (and ``sin``). Are there some specific reasons for including this one?

Comment (leosartaj) [inline sympy/functions/elementary/trigonometric.py]: No nothing specific, I found it interesting. If not as a rewrite, then what is the other way to handle it?

Comment (jksuom) [inline sympy/functions/elementary/trigonometric.py]: Perhaps the formula should not belong to this PR at all. There are product formulas for many other functions, but they have not been included in SymPy. They probably do not fit well into its design.

**Comment by leosartaj** [comment]
I have removed the Product rewrite.

**Comment by leosartaj** [comment]
Is this good to go? (after squashing)

**Comment by jksuom** [comment]
I think this can be squashed. It looks ok.

**Comment by leosartaj** [comment]
I have squashed the commits @jksuom.
